### ISSUE TITLE
color-picker@fmete - Stuck when trying to pick colors 

### ISSUE THREAD

**Issue by Siphalor** [linuxmint/cinnamon-spices-applets#4331]
```
 * Applet version/Build date: 2021-12-11 11:22:30 (last-edited: 1639218150)
 * Cinnamon version: 5.2.6
 * Distribution: Linux Mint 20.3 Cinnamon
 * Graphics hardware *and* driver used: Intel Corporation UHD Graphics 620 [8086:5917] with Mesa Intel(R) UHD Graphics 620 (KBL GT2)
 * Snippet or attachment of ~/.cinnamon/glass.log: N/A
 * Snippet or attachment of ~/.xsession-errors: No logs related to this applet
 ```

**Notify author of applet:** @icarter09 

**Issue**

When triggering the color picking functionality, through either a click or the keybinding, the cursor will transform into a crosshair, but neither clicking nor pressing escape will confirm or exit the process.

The cursor is stuck like this until Cinnamon is restarted.

All clicks over applications are captured by the applet and not handed to the application.
Clicks on panels work like normal, this makes navigation and restarting Cinnamon easily possible.

**Steps to reproduce**

1. Click on the icon / Use your keybinding
2. Try to click anywhere on the screen to pick the color (nothing happens)
3. Try to press escape to leave the picking mode (nothing happens)

**Expected behaviour**

Clicking anywhere on the screen should pick the color and (per the config) open a popup.

Alternatively, pressing escape should leave the picking mode and should return the cursor back to normal.

**Other information**

<details><summary>Configuration</summary>

```json
{
    "head": {
        "type": "header",
        "description": "Color Pick Settings"
    },
    "combo-selection": {
        "type": "combobox",
        "default": "1",
        "description": "Color code format",
        "options": {
            "HEX-Color: #xxxxxx": "1",
            "HEX-Color: #XXXXXX": "2",
            "RGB-Color: 250,250,250": "3",
            "RGB-Color: (250,250,250)": "4"
        },
        "tooltip": "Select the format of the color code, which will be copied to the clipboard by the color picker.",
        "value": "1"
    },
    "div1": {
        "type": "separator"
    },
    "head1": {
        "type": "header",
        "description": "Applet Settings"
    },
    "pick-notification": {
        "type": "checkbox",
        "default": "true",
        "description": "Show picked color code in a popup notification",
        "tooltip": "Show the picked color code, which has been copied to the clipboard, in a popup notification.",
        "value": "true"
    },
    "icon-name": {
        "type": "iconfilechooser",
        "default": "",
        "description": "The icon to use in the panel",
        "tooltip": "Select an icon file, or you can type an icon name into the entry box",
        "value": ""
    },
    "keybinding-test": {
        "type": "keybinding",
        "description": "Keybinding, which activates the color picker",
        "default": "<Primary><Alt>p",
        "value": "<Primary><Alt>p"
    },
    "__md5__": "45992efe29f2a20420a72c97f6364e01"
}
```

</details>

**Comment by fredcw** [comment]
@Siphalor I just tested on Mint 20.3 and it seems to be working ok for me so not sure what's happening. btw, shouldn't Mint 20.3 be cinnamon version 5.2.7

**Comment by Drugwash2** [comment]
**1.** Is the applet installation complete? More precisely: is there a **cp.py** file in **~/.local/share/cinnamon/applets/color-picker@fmete** ?
If not then please remove the applet completely and reinstall it.

**2.** Do you have Python 3.x installed on your system? Is the installation functional? (check by typing `python3 --version` in Terminal)
If not then please (re)install the Python 3 package. Do not attempt to uninstall first as that would most certainly break many things.

**3.** The applet requires certain prerequisites and while it will notify about the lack of most of them it assumes some would exist by default. Therefore please check by whatever means comfortable to you - Terminal or Synaptic - whether you have the following packages installed:

- **apturl** (requires **apturl-common**)
- **python3-xlib**
- **xclip**

If any of them is missing please take note which one(s), install them, and then report back what was missing, what you installed and whether that fixed the issue.

If it still wasn't fixed then you may have to track the problem down by editing applet files adding some debug info or breakpoints to see where it gets stuck.
First one would be **cp.py**; adding `sys.exit()` in **waitClick()** right before `while 1:` should immediately return from the Python script without the need for clicking anywhere and everything should be back to normal (no color copied to clipboard though).
Let's hope we won't get there though.

**Comment by Siphalor** [comment]
**TL;DR:** I got it fixed, it was about multiple python installations, see last heading for more information.

Ok, let's see what we got :)

> I just tested on Mint 20.3 and it seems to be working ok for me so not sure what's happening. btw, shouldn't Mint 20.3 be cinnamon version 5.2.7

You are indeed correct, it reports cinnamon version 5.2.7 now. I could swear that I copied it from `cinnamon --version` though.

Now about the closer debugging:

**1.** The installation seems correct, there's the `cp.py` file and everything else in the applet directory.

**2.**I actually have two Python 3 installations (details below), both have `python-xlib` installed and seem to work (not sure which is used by the applet):
- **a linuxbrew installation** with `python@3.9.12` and `python-xlib@0.31-py2.py3-none-any` via pip. Accessible via `python3`
- **the "system installation"** with apt packages `python3.8@3.8.10-0ubuntu1~20.04.4` and `python3-xlib@0.5.0ubuntu4`. Accessible via `python` or `/usr/bin/python3`

**3.** All the required packages are installed (see above for the Python situation).

### Cause and solution

Since `color-picker@fmete` calls `python3` it picks whatever version comes first in the PATH. This is the linuxbrew version for me.

Through manually calling the python script I found out that the `gi` module was missing in the linuxbrew installation.
I was able to fix that by installing the `gobject` and `PyGObject` pip packages for that installation.

I'm honestly not sure if my setup is edge case enough to just ignore it or if the applet should just use `/usr/bin/python3` or make it configurable.

**Comment by Drugwash2** [comment]
Usually the Python bits found in X-lets are there only because there was no JS method to perform what the author needed, or it was way easier through Python than through JS. Found myself in such situation twice already.

To be honest I don't think there is any (easy) way for an applet to select precisely what version of Python to use, other than picking a specific version number - such as `python3.6` or `python3.9` - but in that case tough luck if the user doesn't have that particular Python version installed. Such thing would create massive incompatibilies.

I believe historically the calling of `python` vs `python3` was meant for distinction between Python 2.x and Python 3.x; at least that's how it works on my Mint 19.2 system. If you have two distinct Python**3** versions called by `python` and `python3` respectively that might open way to further potential issues in applications that expect Python2 on a `python` call. Someone please correct me if I'm wrong.

To somehow mitigate two different versions of the same application/library/etc one can set up **update-alternatives**, but that may present certain limitations itself.

Anyway, glad you found the culprit and fixed the issue.

**Comment by Siphalor** [comment]
The issue in my case is not actually `python` vs `python3`, I'm sorry if I failed to communicate the issue.

The applet is actually expecting that `python3` is the apt installation.
I assume this because the code assumes that `python3-gi` is installed, and asks the user to install `python3-xlib` if that's not already installed:

https://github.com/linuxmint/cinnamon-spices-applets/blob/28cc9ff6e3ef6093ccfb31caee37d84921b81832/color-picker%40fmete/files/color-picker%40fmete/applet.js#L148-L150

In my current setup I have a second python installation that comes before `/usr/bin/` in the `PATH` and therefore gets executed because the code just calls `python3` without a path:

https://github.com/linuxmint/cinnamon-spices-applets/blob/28cc9ff6e3ef6093ccfb31caee37d84921b81832/color-picker%40fmete/files/color-picker%40fmete/applet.js#L143

Arguably, this could be fixed by changing this call to `/usr/bin/python3` which I think more reliably points to the apt installation.

This is also already done for `xclip` a few lines above. I'd really like to know the reasoning why this is done for `xclip`, but nor for `python3`.

https://github.com/linuxmint/cinnamon-spices-applets/blob/28cc9ff6e3ef6093ccfb31caee37d84921b81832/color-picker%40fmete/files/color-picker%40fmete/applet.js#L140

It should be noted though, that generally in this repository calls to python are done with just `python3`, whereas the actual python scripts have `#!/usr/bin/python3` as their shebang instead of `#!/usr/bin/env python3`.

Whatever the final decision on this will be, it would be nice to have some kind of error message in these cases instead of just getting stuck in the middle of the process.
Arguably, `python3-gi` should also be added to the required dependencies.

**Comment by Drugwash2** [comment]
Not your fault, the mentioning of `python` vs `python3` was just a side note unrelated to the current issue.

The usage of `/usr/bin/python3` would indeed be the correct way to call the interpreter. Most likely part of the xlets code here dates back when there barely was a single Python3 installation on any given system so there wasn't much point to specify the full path. Guess progress bites back.

For the dependencies I don't know if there's much to do other than gracefully bail out when missing, by using `try/except`. So far I haven't seen any way to notify the user from this kind of Python call, but at least theoretically there could be a way to access the notifications system. The most I've seen were `print()` commands but those are useless unless the code is called through Terminal, which does not happen in the case of xlets. Even piping error messages through Looking Glass would be of some help. Hopefully someone would figure out how to do either.

**Comment by SebastJava** [comment]
```
Applet version/Build date: Updated on 2023-02-14 03:54:44
Cinnamon version: 5.6.8
Distribution: Linux Mint 21.1 Vera
Graphics:
  Device-1: Intel HD Graphics 530 driver: i915 v: kernel
  Display: x11 server: X.Org v: 1.21.1.3 driver: X: loaded: modesetting
    unloaded: fbdev,vesa gpu: i915 resolution: 1920x1080~60Hz
  OpenGL: renderer: Mesa Intel HD Graphics 530 (SKL GT2) v: 4.6 Mesa 22.2.5
```

I am facing this issue too. With a little annoying twist: It gets stuck only sometimes, not every time.

I tried to completely uninstall and reinstall this applet, but it did not solve my problem. I have been facing this issue for quite a long time before I finally made the effort to come here and post this. Sorry, I haven't tried all your proposed solutions yet. Maybe later.

**Comment by SebastJava** [comment]
@icarter09
I removed it. It was bothering me because of this bug described here. I hope this could get fixed. It was a great instant applet, before it started to randomly freeze from time to time. :(

**Comment by Siphalor** [comment]
@SebastJava
Until my PR gets merged (currently not having high hopes), there are basically three possible workarounds:

1. An easy workaround but not persistent if you regularly update your custom python installation:
	Just install `xlib` in your main python installation (`python3 -m pip install xlib`)
1. An easy workaround but not persistent over spice updates:
	- Go to `~/.local/share/cinnamon/applets/color-picker@fmete`
	- Open `applet.js` and replace `python3` in line 143 with `/usr/bin/python3`
1. Ensure that your system python (`/usr/bin/python3`) is the main python executable.
	You can do that by:
	- Uninstalling any other python installations, probably not desired
	- Putting the bin folders of other python installations later in the `PATH` variable, might not be easily possible if the setup is automated

**Comment by SebastJava** [comment]
@Siphalor Thanks for the quick and rich reply. I am trying your option 2.

This is a "randomly" occurring "freeze". So it's only after several days of bug-free activity that I will be able to reasonably conclude that the problem is really fixed. Until the next spice update, of course.

**Comment by SebastJava** [comment]
@Siphalor wrote:
>     2. An easy workaround but not persistent over spice updates:
>        
>        * Go to `~/.local/share/cinnamon/applets/color-picker@fmete`
>        * Open `applet.js` and replace `python3` in line 143 with `/usr/bin/python3`

That didn't solve my issue. I checked: this edit in line 143 is still there.

I have read the OP here again, and my issue is the same, apparently, with just one difference: this bug only occurs like approx. 10% of the time or so.

Trying to figure out what is wrong, i thought maybe my gestures are too erratic. My mouse and mouse buttons are quite sensitive. Sometimes i accidentally make unwanted drag and drops. Or double-clicks. I tried to deliberately do such erratic moves. This color-picker still works fine when doing foolish drag-and-drops or double-clicks. So, i don't know what is causing this issue.

I am pretty sure this color-picker was working perfectly for years, and started being "buggy" a few months ago or so. And i haven't changed my mouse since early 2020. It is a wireless Logitech M310 and it works fine on every other application.

**Comment by SebastJava** [comment]
Random testing:
- [x] Uninstall and re-install
- [ ] Configure: show notifications = off

**Comment by SebastJava** [comment]
The panel is still responding when everything else freezes. So, i can just click again on the color-picker icon to reset this color-picker and get back to normal.
- [ ] re-tested and confirmed

**Comment by SebastJava** [comment]
Now on new LM Cinnamon 21.3 fresh install. No problem so far...

**Comment by SebastJava** [comment]
UPDATE: I still get the same bug. Only once in a while. Maybe because I go fast and click again while the previous notification is still displayed?
